File: src/control_algorithms_cpp/src/simple_goto.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp> // <-- CAMBIADO A STAMPED
#include <nav_msgs/msg/odometry.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <cmath>
#include <algorithm>

class SimpleGoTo : public rclcpp::Node
{
public:
    SimpleGoTo() : Node("simple_goto_node")
    {
        // 1. Publicador estricto para Gazebo (TwistStamped)
        cmd_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>("/cmd_vel", 10);
        
        // 2. Suscriptor (QoS en 10, que por defecto es RELIABLE, compatible con Gazebo)
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odometry/filtered", 10, 
            std::bind(&SimpleGoTo::odom_callback, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(50),
            std::bind(&SimpleGoTo::control_loop, this));

        RCLCPP_INFO(this->get_logger(), "🚀 Nodo iniciado. Objetivo: X=%.2f, Y=%.2f", target_x_, target_y_);
    }

private:
    double current_x_ = 0.0;
    double current_y_ = 0.0;
    double current_yaw_ = 0.0;
    bool valid_odom_ = false;

    double target_x_ = 1.0;
    double target_y_ = 1.0;

    double k_linear_ = 0.5;
    double k_angular_ = 1.5;
    double max_v_ = 0.2;
    double max_w_ = 1.0;
    double distance_tolerance_ = 0.05;
    bool goal_reached_ = false;

    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr cmd_pub_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

    void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
        current_x_ = msg->pose.pose.position.x;
        current_y_ = msg->pose.pose.position.y;

        tf2::Quaternion q(
            msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
            msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);
        tf2::Matrix3x3 m(q);
        double roll, pitch, yaw;
        m.getRPY(roll, pitch, yaw);
        current_yaw_ = yaw;
        
        if (!valid_odom_) {
            RCLCPP_INFO(this->get_logger(), "✅ Primera odometría recibida! X=%.2f", current_x_);
            valid_odom_ = true;
        }
    }

    void control_loop()
    {
        if (!valid_odom_ || goal_reached_) {
            return; 
        }

        // CREAMOS EL MENSAJE STAMPED
        auto msg = geometry_msgs::msg::TwistStamped();
        msg.header.stamp = this->now();
        msg.header.frame_id = "base_link";

        double dx = target_x_ - current_x_;
        double dy = target_y_ - current_y_;
        double distance_error = std::sqrt(dx*dx + dy*dy);
        double target_yaw = std::atan2(dy, dx);
        double yaw_error = target_yaw - current_yaw_;

        while (yaw_error > M_PI) yaw_error -= 2 * M_PI;
        while (yaw_error < -M_PI) yaw_error += 2 * M_PI;

        if (distance_error < distance_tolerance_) {
            RCLCPP_INFO(this->get_logger(), "🏁 ¡META ALCANZADA! Distancia: %.3fm", distance_error);
            msg.twist.linear.x = 0.0;
            msg.twist.angular.z = 0.0;
            cmd_pub_->publish(msg);
            goal_reached_ = true;
            return;
        }

        // ACCEDEMOS USANDO msg.twist...
        msg.twist.linear.x = std::min(k_linear_ * distance_error, max_v_);
        msg.twist.angular.z = std::max(std::min(k_angular_ * yaw_error, max_w_), -max_w_);

        // Imprimir logs de control
        static int log_counter = 0;
        if (log_counter++ % 20 == 0) {
            RCLCPP_INFO(this->get_logger(), 
                "Dist: %.2f m | YawErr: %.2f rad | Cmd: [v=%.2f, w=%.2f]", 
                distance_error, yaw_error, msg.twist.linear.x, msg.twist.angular.z);
        }

        cmd_pub_->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimpleGoTo>());
    rclcpp::shutdown();
    return 0;
}

File: simple_goto.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
import math

class SimpleGoTo(Node):
    def __init__(self):
        super().__init__('simple_goto_node')
        
        # Publicador a los motores y suscriptor al EKF
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self.odom_sub = self.create_subscription(Odometry, '/odometry/filtered', self.odom_callback, 10)
        
        # --- LA META ---
        self.target_x = 1.2  # Ir 1 metro hacia adelante
        self.target_y = 0.5  # Sin moverse a los lados
        
        # Estado actual del robot
        self.current_x = 0.0
        self.current_y = 0.0
        self.current_yaw = 0.0
        
        # --- PARAMETROS DE CONTROL (Ajustables) ---
        self.k_linear = 0.6       # Qué tan fuerte acelera al avanzar
        self.k_angular = 0.5      # Qué tan fuerte gira
        self.max_v = 0.2          # Límite de velocidad lineal (m/s)
        self.max_w = 0.4          # Límite de velocidad angular (rad/s)
        
        self.dist_tolerance = 0.1 # Se detiene a 10 cm de la meta
        self.yaw_tolerance = 0.1  # Tolerancia de +- 5 grados para apuntar
        
        self.goal_reached = False

    def get_yaw_from_quaternion(self, q):
        # Conversión manual de Cuaternión a Ángulo Euler (Z)
        siny_cosp = 2 * (q.w * q.z + q.x * q.y)
        cosy_cosp = 1 - 2 * (q.y * q.y + q.z * q.z)
        return math.atan2(siny_cosp, cosy_cosp)

    def odom_callback(self, msg):
        if self.goal_reached:
            return

        # 1. Leer posición actual del EKF
        self.current_x = msg.pose.pose.position.x
        self.current_y = msg.pose.pose.position.y
        self.current_yaw = self.get_yaw_from_quaternion(msg.pose.pose.orientation)
        
        # 2. Calcular errores matemáticos
        dx = self.target_x - self.current_x
        dy = self.target_y - self.current_y
        distance = math.sqrt(dx**2 + dy**2)
        
        target_yaw = math.atan2(dy, dx)
        yaw_error = target_yaw - self.current_yaw
        
        # Normalizar el error del ángulo entre -PI y PI
        while yaw_error > math.pi: yaw_error -= 2 * math.pi
        while yaw_error < -math.pi: yaw_error += 2 * math.pi

        cmd = Twist()

        # 3. Lógica de Control (Máquina de estados simple)
        if distance < self.dist_tolerance:
            self.get_logger().info('🏁 ¡META ALCANZADA! Apagando motores.')
            self.cmd_pub.publish(Twist()) # Mandar ceros
            self.goal_reached = True
            return

        # AUMENTAMOS LA TOLERANCIA a 0.25 rad (~14 grados) para que no sea tan estricto
        self.yaw_tolerance = 0.25 

        # Fase 1: Si estoy mirando muy lejos de la meta, solo giro
        if abs(yaw_error) > self.yaw_tolerance:
            cmd.angular.z = self.k_angular * yaw_error  # <-- SIN EL MENOS
            cmd.angular.z = max(min(cmd.angular.z, self.max_w), -self.max_w)
            
            # Fuerza mínima para vencer la fricción estática de la oruga
            if cmd.angular.z > 0 and cmd.angular.z < 0.15: cmd.angular.z = 0.15
            if cmd.angular.z < 0 and cmd.angular.z > -0.15: cmd.angular.z = -0.15
            
            self.get_logger().info(f'🔄 Apuntando... Error: {yaw_error:.2f} | Comando Z: {cmd.angular.z:.2f}')
            
        # Fase 2: Ya estoy apuntando a la meta, ahora avanzo
        else:
            cmd.linear.x = self.k_linear * distance
            cmd.linear.x = max(min(cmd.linear.x, self.max_v), 0.08) 
            
            # Corregimos el rumbo suavemente mientras avanza
            cmd.angular.z = self.k_angular * yaw_error * 0.8  # <-- SIN EL MENOS
            self.get_logger().info(f'⬆️ Avanzando {distance:.2f}m | Corrigiendo {yaw_error:.2f}rad')

        # 4. Enviar comando a las orugas
        self.cmd_pub.publish(cmd)

def main(args=None):
    rclpy.init(args=args)
    node = SimpleGoTo()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.cmd_pub.publish(Twist()) # Freno de emergencia si haces Ctrl+C
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()